cmake/GoogleCloudCppLibrary.cmake

```
@@ -0,0 +1,197 @@
+               $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}>
+               $<INSTALL_INTERFACE:include>)
+    target_link_libraries(
+        ${library_target}
+        PUBLIC google-cloud-cpp::grpc_utils google-cloud-cpp::common
+               google-cloud-cpp::${library}_protos)
+    google_cloud_cpp_add_common_options(${library_target})
+    set_target_properties(
+        ${library_target}
+        PROPERTIES EXPORT_NAME google-cloud-cpp::${library}
+                   VERSION "${PROJECT_VERSION}"
+                   SOVERSION "${PROJECT_VERSION_MAJOR}")
+    target_compile_options(${library_target}
+                           PUBLIC ${GOOGLE_CLOUD_CPP_EXCEPTIONS_FLAG})
+
+    add_library(google-cloud-cpp::${library} ALIAS ${library_target})
+
+    # Create a header-only library for the mocks. We use a CMake `INTERFACE`
+    # library for these, a regular library would not work on macOS (where the
+    # library needs at least one .o file). Unfortunately INTERFACE libraries are
+    # a bit weird in that they need absolute paths for their sources.
+    file(
+        GLOB relative_mock_files
+        RELATIVE "${CMAKE_CURRENT_SOURCE_DIR}"
+        ${mocks_globs})
+    list(SORT relative_mock_files)
+    set(mock_files)
+    foreach (file IN LISTS relative_mock_files)
+        list(APPEND mock_files "${CMAKE_CURRENT_SOURCE_DIR}/${file}")
+    endforeach ()
+    add_library(${mocks_target} INTERFACE)
+    target_sources(${mocks_target} INTERFACE ${mock_files})
+    target_link_libraries(
+        ${mocks_target} INTERFACE google-cloud-cpp::${library}
+                                  GTest::gmock_main GTest::gmock GTest::gtest)
+    set_target_properties(
+        ${mocks_target} PROPERTIES EXPORT_NAME
+                                   google-cloud-cpp::${library}_mocks)
+    target_include_directories(
+        ${mocks_target}
+        INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>
+                  $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}>
+                  $<INSTALL_INTERFACE:include>)
+    target_compile_options(${mocks_target}
+                           INTERFACE ${GOOGLE_CLOUD_CPP_EXCEPTIONS_FLAG})
+
+    # Get the destination directories based on the GNU recommendations.
+    include(GNUInstallDirs)
+
+    # Export the CMake targets to make it easy to create configuration files.
+    install(
+        EXPORT ${library_target}-targets
+        DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/${library_target}"
+        COMPONENT google_cloud_cpp_development)
+
+    # Install the libraries and headers in the locations determined by
+    # GNUInstallDirs
+    install(
+        TARGETS ${library_target} ${protos_target}
+        EXPORT ${library_target}-targets
+        RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+                COMPONENT google_cloud_cpp_runtime
+        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+                COMPONENT google_cloud_cpp_runtime
+                NAMELINK_SKIP
```

Consider (though maybe it is better to post-pone the change to a future PR)

```suggestion
                NAMELINK_COMPONENT google_cloud_cpp_development
```

---

Yeah, let's wait on this and the above (and any other cleanups) to keep this refactor a little more pure.